### PR TITLE
React homework1

### PR DESCRIPTION
**Pull request by grechintanya**

### CODE REVIEW

File: src/components/Courses/Courses.jsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+			<CourseCard
+				key={i}
+				title={item.title}
+				description={item.description}
+				authors={addAuthors(item.authors, mockedAuthorsList)}
+				duration={pipeDuration(item.duration)}
+				creationDate={dateGenerator(item.creationDate)}
+			/>
+		);
+	});
```
Comment (KitTheProgrammer) [inline src/components/Courses/Courses.jsx]: 1. `return` statement is redundant here
2. It is better to use a function here like this: `const renderCourseList = () => <CourseCard/>`

File: src/components/Courses/Courses.jsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+function Courses() {
+	const [query, setQuery] = useState('');
+	const [courses, setCourses] = useState(mockedCoursesList);
+	const [newCourseForm, setNewCourseForm] = useState(false);
+	const [courseData, setCourseData] = useState({
+		title: '',
+		description: '',
+		authors: [],
+		duration: '',
+	});
```
Comment (KitTheProgrammer) [inline src/components/Courses/Courses.jsx]: It is a bad idea to do it like this. Just create separate useState for each value

File: src/components/Courses/Courses.jsx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+	function searchCourses(e) {
+		e.preventDefault();
+		if (query.length) {
+			const courses = mockedCoursesList.filter((course) => {
+				return (
+					course.title.toLowerCase().includes(query.toLowerCase()) ||
+					course.id.toLowerCase().includes(query.toLowerCase())
+				);
+			});
```
Comment (KitTheProgrammer) [inline src/components/Courses/Courses.jsx]: `return` statement is redundant here

File: src/components/Courses/Courses.jsx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+				id: uuidv4(),
+				title: courseData.title,
+				description: courseData.description,
+				creationDate: new Date().toLocaleDateString().replaceAll('.', '/'),
+				duration: Number(courseData.duration),
+				authors: courseData.authors.map((item) => item.id),
+			};
+			setCourses([...courses, newCourse]);
+			setNewCourseForm((prev) => !prev);
+			mockedCoursesList.push(newCourse);
```
Comment (KitTheProgrammer) [inline src/components/Courses/Courses.jsx]: Mutating global variables like this is bad and can cause unexpected behavior. Create a copy of the course list in an `useState` in this component or its parent. Like this: `const [courses, setCourses] = useState([...mockedCoursesList])`

File: src/components/Courses/Courses.jsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+			console.log(mockedCoursesList);
```
Comment (KitTheProgrammer) [inline src/components/Courses/Courses.jsx]: Don't forget to remove debugging code before committing/pushing

File: src/components/Courses/Courses.jsx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		}
+	}
+
+	return (
+		<div className='courses'>
+			{newCourseForm || (
+				<>
+					<header>
```
Comment (KitTheProgrammer) [inline src/components/Courses/Courses.jsx]: Did you mean `newCourseForm && (...`?

If `newCourseForm` is `true` it will display nothing.

Comment (KitTheProgrammer) [reply src/components/Courses/Courses.jsx]: I think I understood what you were trying to do.
You want to hide the courses when you are creating a new one. (there is no such requirement in the homework description tho)

But, there is a better way to do it, just use the ternar operator:
`newCourseForm ? <show list> : <show form>`

File: src/components/Courses/Courses.jsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+					<div className='cards'>{courseList}</div>
+				</>
+			)}
+			{newCourseForm && (
+				<CreateCourse
+					courseData={courseData}
+					setCourseData={setCourseData}
+					handleSubmit={saveNewCourse}
+				/>
+			)}
```
Comment (KitTheProgrammer) [inline src/components/Courses/Courses.jsx]: Highlighting this part as it is a part of my previous comment

File: src/components/Courses/Courses.jsx [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+			const newCourse = {
+				id: uuidv4(),
+				title: courseData.title,
+				description: courseData.description,
+				creationDate: new Date().toLocaleDateString().replaceAll('.', '/'),
+				duration: Number(courseData.duration),
+				authors: courseData.authors.map((item) => item.id),
+			};
+			setCourses([...courses, newCourse]);
+			setNewCourseForm((prev) => !prev);
```
Comment (KitTheProgrammer) [inline src/components/Courses/Courses.jsx]: you can do it like this: `setNewCourseForm(!newCourseForm)`

P.S.
This approach is just not necessary here you don't have any async code

File: src/components/Courses/Courses.jsx [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+					<header>
+						<SearchBar
+							query={query}
+							setQuery={setQuery}
+							handleSubmit={searchCourses}
+						/>
+						<Button
+							buttonText='Add new course'
+							className='btn_add'
+							handleClick={() => setNewCourseForm((prev) => !prev)}
```
Comment (KitTheProgrammer) [inline src/components/Courses/Courses.jsx]: `setNewCourseForm(!newCourseForm)`

File: src/components/CreateCourse/CreateCourse.jsx [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+							],
+						});
+						setAuthors((prev) => [
+							...prev.filter((item) => !(item.id === author.id)),
+						]);
+					}}
+				/>
+			</div>
+		);
+	});
```
Comment (KitTheProgrammer) [inline src/components/CreateCourse/CreateCourse.jsx]: Do it as a render function (I think I described similar issue somewhere higher)

File: src/components/CreateCourse/CreateCourse.jsx [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+import { mockedAuthorsList } from '../../constants';
+import pipeDuration from '../../helpers/pipeDuration';
+import './createCourse.css';
+
+function CreateCourse({ courseData, setCourseData, handleSubmit }) {
+	const [authorName, setAuthorName] = useState('');
+
+	const [authors, setAuthors] = useState(mockedAuthorsList);
+	const authorsList = authors.map((author) => {
+		return (
```
Comment (KitTheProgrammer) [inline src/components/CreateCourse/CreateCourse.jsx]: `return` statement is redundant here

File: src/components/CreateCourse/CreateCourse.jsx [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+			<div className='author' key={author.id}>
+				<p>{author.name}</p>
+				<Button
+					buttonText='Add author'
+					type='button'
+					handleClick={() => {
+						setCourseData({
+							...courseData,
```
Comment (KitTheProgrammer) [inline src/components/CreateCourse/CreateCourse.jsx]: Move this handler to it's own function

File: src/components/CreateCourse/CreateCourse.jsx [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+							authors: [
+								...courseData.authors,
+								{ id: author.id, name: author.name },
+							],
+						});
+						setAuthors((prev) => [
+							...prev.filter((item) => !(item.id === author.id)),
+						]);
```
Comment (KitTheProgrammer) [inline src/components/CreateCourse/CreateCourse.jsx]: You have no need to use prev value here. just use `authors`

Comment (KitTheProgrammer) [inline src/components/CreateCourse/CreateCourse.jsx]: Use a render function for this purpose (I think I already provided an example for similar case somewhere earlier)

File: src/components/CreateCourse/CreateCourse.jsx [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+									(item) => !(item.id === author.id)
+								),
+							],
+						});
+						setAuthors([...authors, author]);
+					}}
+				/>
+			</div>
+		);
+	});
```
Comment (KitTheProgrammer) [inline src/components/CreateCourse/CreateCourse.jsx]: same here

**Review by KitTheProgrammer — CHANGES_REQUESTED** [review]
I checked your app and here is what I found:
1. You can add new authors with empty names
2. You can't cancel the new course creation

File: src/components/Courses/Courses.jsx (hunk 3 above)
Comment (grechintanya) [reply src/components/Courses/Courses.jsx]: Я не дуже зрозуміла, що тут потрібно змінити. Прибрати return?
Без нього фільтр не хоче працювати, а еслінт свариться і кажу, що у колбек-функції має бути return statment

Comment (KitTheProgrammer) [reply src/components/Courses/Courses.jsx]:
```
const courses = mockedCoursesList.filter((course) => 
      course.title.toLowerCase().includes(query.toLowerCase()) ||
      course.id.toLowerCase().includes(query.toLowerCase()));
```

**Review by KitTheProgrammer — APPROVED** [review]
Well done!
20

**Pull request merged**